labs/b9.md

```diff
@@ -0,0 +1,178 @@
+- You should be able to view the web application from your host machine by visiting <url_of_host_machine>:5000, assuming you don’t have anything else running on that port.      
+ 
+ ##### Questions
+ 1. While your three containerized Apache web servers are running in detached mode, paste the output of `docker ps`.
+ 2. Observe that in the output of `docker ps` each container has an associated container ID. Explain why containers have IDs rather than being named after the application, for example httpd. 
+ 3. Now go ahead and stop your containers (see slide 30). Paste the command you used to stop one of the containers. 
+
+Congratulations! You’ve successfully Dockerized and ran a web application with vastly different requirements and configuration than your host machine, without affecting your setup on your machine :) There’s a lot more about Docker and containers to learn about, but I hope this was enough to wrap your head around the basic concepts and get some experience working with it.        
+
+For further reading, I recommend just reading the [official documentation](https://docs.docker.com/) so you can see what is possible with the Docker container format.    
+
+#### Puppet
+
+To get some hands on experience with Puppet, we are going to write a very basic manifest that pulls quotes from the internet and appends them to a file every two minutes. Although you were taught that in production environments we have the puppet master running on its own server, for simplicity you will apply the manifest locally, making you both the puppet master and the puppet agent. If you are confused about the Puppet vocabulary, check out slide 21.
+
+First things first. We plan to pull quotes from the web, so we need to ensure the `wget` package is installed on our system. Next we need to create a user, 'quotes', that runs this command for us. Take note of the dependency between the 'quotes' user and the 'quotegather' group existing. 
+
+Because we have full control over our user, let's make its home directory `/tmp`. Since this user has one purpose, to pull quotes from the web, they don't need a login shell. Go ahead and set it to `/bin/false`.
+
+Now we need a `cron` resource that grabs the quotes for us. I went ahead and filled in the command, so all you need to do is specify which user is to run the cron job and the interval at which it runs (every 2 minutes).
+
+Below I've provided a skeleton manifest `quotes.pp`. This can also be found in `~/decal-labs/b9/puppet/quotes.pp`, but comments are included here for your convenience.
+
+```puppet
+package { '': } # What package is required for this script to run?
+
+group { 'quotegather':
+  ensure  => '', # Do we want the group on this system?
+}
+
+user { 'quotes':
+  ensure  => '', # Do we want the user on this system?
+  gid     => , # Note that we can pass a group name to gid
```

Comment:
I feel like just having a comma here is a bit confusing, not sure how to make it better though

Reply:
I took brian's suggestion and linked an example manifest that has `gid` in it. that should clear this up.